asdfghjkl/vector.py

```diff
@@ -0,0 +1,131 @@
+        elif isinstance(vectors, list):
+            for p, v in zip(params, vectors):
+                assert p.shape == v.shape
+                self.vectors[p] = v
+        elif isinstance(vectors, dict):
+            for p in params:
+                assert p.shape == vectors[p].shape
+            self.vectors = vectors
+        else:
+            raise TypeError(f'Invalid vectors type: {type(vectors)}')
+
+    def values(self):
+        return self.vectors.values()
+
+    def __add__(self, other):
+        vectors = [v1 + v2 for v1, v2 in zip(self.values(), other.values())]
+        return ParamVector(self.params, vectors)
+
+    def __iadd__(self, other):
+        for v1, v2 in zip(self.values(), other.values()):
+            v1 += v2
+
+    def add(self, other, alpha=1):
+        vectors = [v1.add(v2, alpha=alpha) for v1, v2 in zip(self.values(), other.values())]
+        return ParamVector(self.params, vectors)
+
+    def add_(self, other, alpha=1):
+        for v1, v2 in zip(self.values(), other.values()):
+            v1.add_(v2, alpha=alpha)
+        return self
+
+    def extend(self, other):
+        self.params.extend(other.params)
+        self.vectors.update(other.vectors)
+
+    def mul(self, value):
+        return ParamVector(self.params, [v.mul(value) for v in self.values()])
+
+    def mul_(self, value):
+        for key in self.vectors:
+            self.vectors[key].mul_(value)
+        return self
+
+    def dot(self, other):
+        return torch.dot(self.get_flatten_vector(), other.get_flatten_vector())
+
+    def norm(self):
+        return torch.norm(self.get_flatten_vector())
+
+    def get_vectors_by_module(self, module: nn.Module):
+        params = [p for p in module.parameters()]
+        return self.get_vectors_by_params(params)
+
+    def get_vectors_by_params(self, params: List[torch.Tensor]):
+        vectors = {p: self.vectors[p] for p in params if p in self.vectors}
+        if len(vectors) == 0:
+            return None
+        params = list(vectors.keys())
+        return ParamVector(params, vectors)
+
+    def get_vector_by_param(self, param: torch.Tensor, default=None) -> torch.Tensor:
+        return self.vectors.get(param, default)
+
+    def get_flatten_vector(self):
+        flat_vecs = [v.flatten() for v in self.values()]
+        return torch.cat(flat_vecs)
+
+    def numel(self):
+        return sum(v.numel() for v in self.values())
+
+    def copy(self):
+        return ParamVector(self.params, copy.deepcopy(list(self.values())))
+
+
+def reduce_vectors(vectors: ParamVector, is_master=True, all_reduce=False) -> ParamVector:
+    # pack
+    packed_tensor = vectors.get_flatten_vector()
+    if all_reduce:
+        # all-reduce
+        dist.all_reduce(packed_tensor)
+    else:
+        dist.reduce(packed_tensor, dst=0)
+    if all_reduce or is_master:
+        # unpack
+        rst = ParamVector(vectors.params, packed_tensor)
+    else:
+        rst = None
+
+    dist.barrier()
```

Comment:
Does `all_reduce` have the implicit effect as `barrier`?

Reply:
It might have. But I think it is safe to put a barrier. This barrier is also for `reduce` (not only `all_reduce`).